web/source/text/processor.ts

```diff
@@ -30,49 +30,92 @@ namespace com.keyman.text {
       let keyman = com.keyman.singleton;
 
+      let outputTarget = Lkc.Ltarg;
+      let preInput = Mock.from(outputTarget);
+      let ruleBehavior = new RuleBehavior();
+
       // check if exact match to SHIFT's code.  Only the 'default' and 'shift' layers should have default key outputs.
       if (keyShiftState == Codes.modifierCodes['SHIFT']) {
         keyShiftState = 1; // It's used as an index in some called methods.
       }
 
-      // If this was triggered by the OSK -or- if it was triggered by a 'synthetic' OutputTarget (TouchAlias, Mock)
-      // that lacks default key processing behavior.
+      let matched = false;
+      var char = '';
+      var special: EmulationKeystrokes;
       if(usingOSK || outputTarget.isSynthetic) {
-        let result = DefaultOutput.applyCommand(Lkc, keyShiftState);
-
-        if(result) {
-          return result;
+        matched = true;  // All the conditions below result in matches until the final else, which restores the expected default
+                         // if no match occurs.
+
+        if(DefaultOutput.isCommand(Lkc)) {
+          // Note this in the rule behavior, return successfully.  We'll consider applying it later.
+          ruleBehavior.triggersDefaultCommand = true;
+
+          // We'd rather let the browser handle these keys, but we're using emulated keystrokes, forcing KMW
+          // to emulate default behavior here.
+        } else if(special = DefaultOutput.forSpecialEmulation(Lkc)) { 
+          switch(special) {
+            case EmulationKeystrokes.Backspace:
+              keyman.interface.defaultBackspace(outputTarget);
+              break;
+            case EmulationKeystrokes.Enter:
+              outputTarget.handleNewlineAtCaret();
+              break;
+            case EmulationKeystrokes.Space:
+              keyman.interface.output(0, outputTarget, ' ');
+              break;
+            // case '\u007f': // K_DEL
+              // // For (possible) future implementation.
+              // // Would recommend (conceptually) equaling K_RIGHT + K_BKSP, the former of which would technically be a 'command'.
+            default:
+              // In case we extend the allowed set, but forget to implement its handling case above.
+              ruleBehavior.errorLog = "Unexpected 'special emulation' character (\\u" + (special as String).kmwCharCodeAt(0).toString(16) + ") went unhandled!";
+          } 
+        } else {
+          // Back to the standard default, pending normal matching.
+          matched = false;
         }
-      } else if(Lkc.Lcode == 8) {
-        keyman.interface.defaultBackspace();
-        return '';
       }
 
-      let code: string;
+      if(!matched) {
+        if(char = DefaultOutput.forAny(Lkc, keyShiftState)) {
+          special = DefaultOutput.forSpecialEmulation(Lkc)
+          if(special == EmulationKeystrokes.Backspace) {
+            // A browser's default backspace may fail to delete both parts of an SMP character.
```

Comment:
This catches the same case as the previous comment while preventing all specially-handled cases from being omitted when not in a 'synthetic' mode.